pkg/k8s/events/types.go

```diff
@@ -30,7 +42,141 @@ const (
+	return fmt.Sprintf("%s-%s", m.Kind, m.Type)
+}
+
+// EventType represents the type of event we have received from Kubernetes: added, updated, or deleted
+type EventType string
+
+const (
+	// Added is an EventType representing a new object being added.
+	Added EventType = "added"
+
+	// Updated is an EventType representing an object being updated.
+	Updated EventType = "updated"
+
+	// Deleted is an EventType represent an object being deleted.
+	Deleted EventType = "deleted"
+)
+
+// Kind is used to record the kind of announcement
+type Kind string
+
+// String returns the string representation of the Kind
+func (at Kind) String() string {
+	return string(at)
+}
+
+// Added returns the string representation of the Kind combined with the added keyword to be used for a PubSub Topic.
+func (at Kind) Added() string {
+	return fmt.Sprintf("%s-%s", at, Added)
+}
+
+// Updated returns the string representation of the Kind combined with the updated keyword to be used for a PubSub Topic.
+func (at Kind) Updated() string {
+	return fmt.Sprintf("%s-%s", at, Updated)
+}
+
+// Deleted returns the string representation of the Kind combined with the deleted keyword to be used for a PubSub Topic.
+func (at Kind) Deleted() string {
+	return fmt.Sprintf("%s-%s", at, Deleted)
+}
+
+const (
+	// Pod is a Kind for Kubernetes pod events.
+	Pod Kind = "pod"
+
+	// Endpoint is the Kind for Kubernetes events.
+	Endpoint Kind = "endpoint"
+
+	// Namespace is the Kind for Kubernetes namespace events.
+	Namespace Kind = "namespace"
+
+	// Service is the Kind for Kubernetes service events.
+	Service Kind = "service"
+
+	// ServiceAccount is the Kind for Kubernetes service account events.
+	ServiceAccount Kind = "serviceaccount"
+
+	// TrafficSplit is the Kind for Kubernetes traffic split events.
+	TrafficSplit Kind = "trafficsplit"
+
+	// RouteGroup is the Kind for Kubernetes route group events.
+	RouteGroup Kind = "routegroup"
+
+	// TCPRoute is the Kind for Kubernetes tcp route events.
+	TCPRoute Kind = "tcproute"
+
+	// TrafficTarget is the Kind for Kubernetes traffic target events.
+	TrafficTarget Kind = "traffictarget"
+
+	// Ingress is the Kind for Kubernetes ingress events.
+	Ingress Kind = "ingress"
+
+	// MeshConfig is the Kind for Kubernetes meshconfig events.
+	MeshConfig Kind = "meshconfig"
+
+	// MeshRootCertificate is the Kind for Kubernetes mrc events.
+	MeshRootCertificate Kind = "meshrootcertificate"
+
+	// Egress is the Kind for Kubernetes egress events.
+	Egress Kind = "egress"
+
+	// IngressBackend is the Kind for Kubernetes ingress backend events.
+	IngressBackend Kind = "ingressbackend"
+
+	// RetryPolicy is the Kind for Kubernetes retry policy events.
+	RetryPolicy Kind = "retry"
+
+	// UpstreamTrafficSetting is the Kind for Kubernetes updstream traffic settings events.
+	UpstreamTrafficSetting Kind = "upstreamtrafficsetting"
+)
+
+// GetKind returns the Kind for the given k8s object.
+func GetKind(obj interface{}) Kind {
+	switch obj.(type) {
```

Comment:
Ya I had that originally but in unit tests it doesn’t get the fields populated (even if written and retrieved from a fake client)

would prefer that though open to alternatives

Reply:
The kind would need to be set on the object explicitly in the test since I think that's only set when (de)serializing. We could roll our own implementation of runtime.Object for this test too. I don't see any other tests failing with this change.

Reply:
that leaves us vulnerable to typo's though, where if I put in the type and it doesn't match the actual type that k8s will populate.

Whereas this approach a typo wouldn't matter since we are controlling it. I guess we could have an E2E test that validates all the types, but that feels like the same approach as current.

Reply:
I ran into this along time ago, I guess it still isn't working?  

This was a possible alternative I had found but didn't use.  Not sure if it would work (or is still relevant but thought I might mention: https://github.com/Azure/azure-k8s-metrics-adapter/blob/1b25752ec2eac4e9c484cda06b9aae4a4a6af1db/pkg/controller/controller.go#L172-L175

Reply:
Looks like that code doesn't exist anymore :/ I'd probably opt to not go with that. 

It looks like it might not work with custom types as well, due to leveraging the default schemes